drivers/Aqara/aqara-presence-sensor/src/discovery.lua

```diff
@@ -0,0 +1,92 @@
+local log = require "log"
+local discovery = {}
+local fields = require "fields"
+local discovery_mdns = require "discovery_mdns"
+local socket = require "cosock.socket"
+
+-- mapping from device DNI to info needed at discovery/init time
+local device_discovery_cache = {}
+
+local function set_device_field(driver, device)
+  local device_cache_value = device_discovery_cache[device.device_network_id]
+
+  -- persistent fields
+  device:set_field(fields.DEVICE_IPV4, device_cache_value.ip, { persist = true })
+  device:set_field(fields.DEVICE_INFO, device_cache_value.device_info, { persist = true })
+  device:set_field(fields.CREDENTIAL, device_cache_value.credential, { persist = true })
+end
+
+local function update_device_discovery_cache(driver, dni, ip, credential)
+  local device_info = driver.discovery_helper.get_device_info(driver, dni, ip)
+  device_discovery_cache[dni] = {
+    ip = ip,
+    device_info = device_info,
+    credential = credential,
+  }
+end
+
+local function try_add_device(driver, device_dni, device_ip)
+  log.trace(string.format("try_add_device : dni= %s, ip= %s", device_dni, device_ip))
+
+  local credential = driver.discovery_helper.get_credential(driver, device_dni, device_ip)
+
+  if not credential then
+    log.error(string.format("failed to get credential. dni= %s, ip= %s", device_dni, device_ip))
+    return
+  end
+
+  update_device_discovery_cache(driver, device_dni, device_ip, credential)
+  local create_device_msg = driver.discovery_helper.get_device_create_msg(driver, device_dni, device_ip)
+  driver:try_create_device(create_device_msg)
+end
+
+function discovery.device_added(driver, device)
+  set_device_field(driver, device)
+  device_discovery_cache[device.device_network_id] = nil
+  driver.lifecycle_handlers.init(driver, device)
```

Comment:
I don't believe you should be calling lifecycle handlers directly, this will be called by the framework

Reply:
It seems like this is needed in order to make sure init is called after added. Init requires the device fields to have been populated.